data/transactions/logic/eval.go

```diff
@@ -4764,56 +4766,45 @@ func opBase64Decode(cx *EvalContext) error {
 	cx.stack[last].Bytes = bytes
 	return nil
 }
-func hasDuplicateKeys(jsonText []byte) (bool, map[string]json.RawMessage, error) {
+
+func isPrimitiveJSON(jsonText []byte) (bool, error) {
 	dec := json.NewDecoder(bytes.NewReader(jsonText))
-	parsed := make(map[string]json.RawMessage)
 	t, err := dec.Token()
 	if err != nil {
-		return false, nil, err
+		return false, err
 	}
 	t, ok := t.(json.Delim)
 	if !ok || t.(json.Delim).String() != "{" {
-		return false, nil, fmt.Errorf("only json object is allowed")
-	}
-	for dec.More() {
-		var value json.RawMessage
-		// get JSON key
-		key, err := dec.Token()
-		if err != nil {
-			return false, nil, err
-		}
-		// end of json
-		if key == '}' {
-			break
-		}
-		// decode value
-		err = dec.Decode(&value)
-		if err != nil {
-			return false, nil, err
-		}
-		// check for duplicates
-		if _, ok := parsed[key.(string)]; ok {
-			return true, nil, nil
-		}
-		parsed[key.(string)] = value
+		return true, nil
 	}
-	return false, parsed, nil
+	return false, nil
 }
 
 func parseJSON(jsonText []byte) (map[string]json.RawMessage, error) {
-	if !json.Valid(jsonText) {
-		return nil, fmt.Errorf("invalid json text")
-	}
-	// parse json text and check for duplicate keys
-	hasDuplicates, parsed, err := hasDuplicateKeys(jsonText)
-	if hasDuplicates {
-		return nil, fmt.Errorf("invalid json text, duplicate keys not allowed")
+	// guard against primitives
+	isPrimitive, err := isPrimitiveJSON(jsonText)
+	if err != nil {
+		return nil, err
+	} else if isPrimitive {
+		return nil, fmt.Errorf("invalid json text, only json object is allowed")
 	}
+
+	// parse JSON with Algorand's standard JSON library
+	var parsed map[string]json.RawMessage
+	err = protocol.DecodeJSON(jsonText, &parsed)
```

Comment:
Also, if we're checking for primitives, I'm surprised we're not checking to see if it's an array.

Reply:
We do check to see if it's an array. I will change the name to reduce confusion.

Reply:
So what happens if "null" is given to protocol.Decode? Surely it doesn't return anything that can possibly be indexed, so we would still fail rather than ever successfully return a value.

Reply:
So there's no way to distinguish "null" from "{}"?

Even if we can't, I still don't think we have to detect the "null".  Every time we parse, we trying to get a field.  If the assignment of null happens as an empty map, when we fetch the field, we'll certainly fail.

Reply:
I don't mind just saying "key not found".

If it's really important to you, you could do the test *after* you find the key is missing.  that way you never do it except in the error path.

Reply:
Yeah that's fair. I'll change it to only check in the case of an error.